test/pipelines/unit/test_pipeline.py

```diff
@@ -24,3 +24,300 @@ def test_max_loops():
+        "add_default": add_default,
+    }
+    data = {
+        "metadata": {"test": "test"},
+        "max_loops_allowed": 100,
+        "components": {
+            "add_two": {},
+            "add_default": {},
+            "double": {
+                "type": "Double",
+                "hash": "789",
+                "init_parameters": {},
+            },
+        },
+        "connections": [
+            {"sender": "add_two.result", "receiver": "double.value"},
+            {"sender": "double.value", "receiver": "add_default.value"},
+        ],
+    }
+    pipe = Pipeline.from_dict(data, components=components)
+    assert pipe.metadata == {"test": "test"}
+    assert pipe.max_loops_allowed == 100
+
+    # Components
+    assert len(pipe.graph.nodes) == 3
+    ## add_two
+    add_two_data = pipe.graph.nodes["add_two"]
+    assert add_two_data["instance"] is add_two
+    assert add_two_data["instance"].add == 2
+    assert add_two_data["input_sockets"] == {
+        "value": InputSocket(name="value", type=int, is_optional=False, sender=None),
+        "add": InputSocket(name="add", type=Optional[int], is_optional=True, sender=None),
+    }
+    assert add_two_data["output_sockets"] == {
+        "result": OutputSocket(name="result", type=int),
+    }
+    assert add_two_data["visits"] == 0
+
+    ## add_default
+    add_default_data = pipe.graph.nodes["add_default"]
+    assert add_default_data["instance"] is add_default
+    assert add_default_data["instance"].add == 1
+    assert add_default_data["input_sockets"] == {
+        "value": InputSocket(name="value", type=int, is_optional=False, sender="double"),
+        "add": InputSocket(name="add", type=Optional[int], is_optional=True, sender=None),
+    }
+    assert add_default_data["output_sockets"] == {
+        "result": OutputSocket(name="result", type=int),
+    }
+    assert add_default_data["visits"] == 0
+
+    ## double
+    double = pipe.graph.nodes["double"]
+    assert double["instance"]
+    assert double["input_sockets"] == {
+        "value": InputSocket(name="value", type=int, is_optional=False, sender="add_two"),
+    }
+    assert double["output_sockets"] == {
+        "value": OutputSocket(name="value", type=int),
+    }
+    assert double["visits"] == 0
+
+    # Connections
+    connections = list(pipe.graph.edges(data=True))
+    assert len(connections) == 2
+    assert connections[0] == (
+        "add_two",
+        "double",
+        {
+            "conn_type": "int",
+            "from_socket": OutputSocket(name="result", type=int),
+            "to_socket": InputSocket(name="value", type=int, is_optional=False, sender="add_two"),
+        },
+    )
+    assert connections[1] == (
+        "double",
+        "add_default",
+        {
+            "conn_type": "int",
+            "from_socket": OutputSocket(name="value", type=int),
+            "to_socket": InputSocket(name="value", type=int, is_optional=False, sender="double"),
+        },
+    )
```

Comment:
Same as above, with an additional check:

```suggestion
    double = Double()
    expected_pipe = Pipeline(metadata={"test": "test"})
    expected_pipe.add_component("add_two", add_two)
    expected_pipe.add_component("add_default", add_default)
    expected_pipe.add_component("double", double)
    expected_pipe.connect("add_two", "double")
    expected_pipe.connect("double", "add_default")
    
    assert pipe == expected_pipe
    assert pipe.get_component("add_two") == expected_pipe.get_component("add_two")
    assert pipe.get_component("add_default") == expected_pipe.get_component("add_default")
```